src/optimizer/rule_impls.cpp

```diff
@@ -386,17 +386,22 @@ void GetToIndexScan::Transform(
       std::vector<oid_t> index_key_column_id_list;
       std::vector<ExpressionType> index_expr_type_list;
       std::vector<type::Value> index_value_list;
-      std::unordered_set<oid_t> index_col_set(
-          index_object->GetKeyAttrs().begin(),
-          index_object->GetKeyAttrs().end());
-      for (size_t offset = 0; offset < key_column_id_list.size(); offset++) {
-        auto col_id = key_column_id_list[offset];
-        if (index_col_set.find(col_id) != index_col_set.end()) {
-          index_key_column_id_list.push_back(col_id);
+
+      // Only pick the index if the query columns match the index's columns in
+      // the same order.
+      auto index_id_list = index_object->GetKeyAttrs();
+      for (size_t offset = 0; (offset < key_column_id_list.size()) &&
+                              (offset < index_id_list.size());
+           offset++) {
+        if (index_id_list[offset] == key_column_id_list[offset]) {
+          index_key_column_id_list.push_back(key_column_id_list[offset]);
           index_expr_type_list.push_back(expr_type_list[offset]);
           index_value_list.push_back(value_list[offset]);
+        } else {
```

Comment:
I don't think we should check for an exact same ordering here. For example, if you have an index on column (a, b), and your predicates are "b = 5 and a = 1", then we should be able to use the index scan. However, the check here won't identify that because it requires the order in the predicates to be exactly the same as in the index.

After thinking about this, I actually think that you should just keep the old `index_key_column_id_list`. You just need to add a flag about whether the lead (highest) column in the index has been referenced in the index. As long as that is true, we should be able to use the index for the scan. Thoughts? @chenboy @vkonagar

Reply:
Agree. I also think we don't need to consider order here. The way to fix this issue is letting the cost model compute the correct cost for these indices.

Reply:
Hi, @chenboy! @vkonagar and I are discussing some implementation details for this on Slack. We've added you into the channel. There's some cost model related issue we think you probably have a better idea on what's going on. Can you take a look at Slack? Thanks!